crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,19 +1,156 @@
+                );
+            }
+            [.., last] => last,
+        };
+
+        let saved_level = f.context().node_level();
+        // Tell the children they need parentheses
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let magic_trailing_comma = USE_MAGIC_TRAILING_COMMA
+            && matches!(
+                find_first_non_trivia_character_after(last.range().end(), f.context().contents()),
+                Some((_, ','))
+            );
+
+        if magic_trailing_comma {
+            // A magic trailing comma forces us to print in expanded mode since we have more than
+            // one element
+            write!(
+                f,
+                [group(&format_args![
+                    // An expanded group always needs parentheses
+                    &text("("),
+                    hard_line_break(),
+                    block_indent(&ExprSequence::new(elts)),
+                    hard_line_break(),
+                    &text(")"),
+                ])]
+            )?;
+        } else if is_parenthesized(*range, elts, f) {
+            // If the tuple has parentheses, keep them. Also only top level tuples are allowed to
+            // elide them
+            write!(
+                f,
+                [group(&format_args![
+                    // If it was previously parenthesized, add them again
+                    &text("("),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    &text(")"),
+                ])]
+            )?;
+        } else {
+            write!(
+                f,
+                [group(&format_args![
+                    // If there were previously no parentheses, add them only if the group breaks
+                    if_group_breaks(&text("(")),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    if_group_breaks(&text(")")),
+                ])]
+            )?;
+        }
+
+        f.context_mut().set_node_level(saved_level);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ExprSequence<'a> {
+    elts: &'a [Expr],
+}
+
+impl<'a> ExprSequence<'a> {
+    const fn new(elts: &'a [Expr]) -> Self {
+        Self { elts }
+    }
+}
+
+impl Format<PyFormatContext<'_>> for ExprSequence<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        for (pos, entry) in self.elts.iter().enumerate() {
+            // We need a trailing comma on the last entry of an expanded group since we have more
+            // than one element
+            if pos == self.elts.len() - 1 {
+                write!(
+                    f,
+                    [
+                        entry.format(),
+                        if_group_breaks(&text(",")),
+                        soft_line_break()
+                    ]
+                )?;
+            } else {
+                write!(f, [entry.format(), text(","), soft_line_break_or_space()])?;
+            }
```

Comment:
You can simplify this by using `joiner`:

```suggestion
		let separator = format_with(|f| write!(f[text(","), soft_line_break_or_space()]));
		f.join_with(separator).entries(self.elts.iter().formatted()).finish()?;
		
		// Write the trailing comma 
		write!(f, [if_group_breaks(&text(","))])?;
```